1937 Adds an indentation preference to the decompiler output

js53867 opened the pull request:
This change also adds support for two indentation styles:

K&R style,

while (x == y) {
    something();
    somethingelse();
}

Allman style,

while (x == y)
{
    something();
    somethingelse();
}

The default style is K&R.

mumbel commented on Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc:
```diff
@@ -2867,6 +2879,33 @@ void PrintC::emitBlockSwitch(const BlockSwitch *bl)
+
+{
+  if (option_indentationStyle == indentation_style_allman) {
+  	if (NULL != indent) {
```
indent is an int4, not pointer, would be better as `if (0 != indent)` or simply `if (indent)`

aerosoul94 commented on Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc:
```diff
@@ -2867,6 +2879,33 @@ void PrintC::emitBlockSwitch(const BlockSwitch *bl)
 }
 
+/// \brief Emits an opening brace according to the indentation style selected in options
+int4 PrintC::emitFormatedStartBrace(int4 indent)
```
Formatted is misspelled.

aerosoul94 commented on Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc:
```diff
@@ -2867,6 +2879,33 @@ void PrintC::emitBlockSwitch(const BlockSwitch *bl)
+    emit->print("{");
+  }
+  else {
+    throw LowlevelError("Unkown indentation style");
```
Unknown is misspelled.

ryanmkurtz commented on Ghidra/Features/Decompiler/src/decompile/cpp/printlanguage.hh:
```diff
@@ -431,6 +431,22 @@ public:
   /// \param nm is the configuration description
   virtual void setCommentStyle(const string &nm)=0;
 
+<<<<<<< HEAD
```
There's a bunch of diff lines that were left in.

mumbel replied:
Did GitHub change their ui? Or is this just odd cause I'm on mobile, though 
 'desktop mode'.  Should probably revert the merge entirely, and handle any debase without pulling in all the commits into this or (which I'm guessing is where the conflict came from, but the website looks different/off)

js53867 replied:
Yeah, this merge has not gone well! Thanks for spotting this, just sorting it out now

js53867 commented:
I'm conscious that the PR rules state that all commits should be squashed into one commit - and that this PR is now four commits having made some suggested changes and fixing merge issues.

What is expected at this point? Should I delete this PR and resubmit as a single commit?

ryanmkurtz commented:
A lot of people don't know how to squash or rebase, so I that rule has been relaxed.  However, I would want to see any bad merge commits squashed out of the final product, even if the result is more than one commit.

js53867 commented:
Sorry, been a while since I last looked at this - I have now updated to remove the previously unintentionally included errors, and squashed back down to one commit.

js53867 commented:
Is there anything which I can do to aid the merging of this PR?

ryanmkurtz commented:
It looks like the conflict-diffs were taken care of, but I'll have to defer reviewing the rest of the PR to our decompiler expert.

js53867 commented:
Thanks Ryan! Will resolve those new conflicts asap

js53867 commented on Ghidra/Features/Decompiler/src/main/java/ghidra/app/decompiler/DecompileOptions.java:
```diff
@@ -784,7 +817,9 @@ public String getXML(DecompInterface iface) {
 		}
+
 		appendOption(buf, "protoeval", protoEvalModel, "", "");
+		appendOption(buf, "indentationstyle", indentationStyle.getOptionString(), "", "");
```
I confess that many of these calls to appendOption are contained within a 
```
if (option != OPTION_OPTIONDEFAULT) {
    ...
}
```

Clause, and I'm both not sure what this protects against, and whether the indentation style check should have a similar check.

js53867 commented:
I believe the conflicts are resolved! Do let me know if there's anything else I can do to help with the review process.

js53867 commented:
Is there anything else that I can do to help with the review process?

XVilka commented:
Now there are new conflicts again. @js53867 we also have this change, I wonder if it makes sense to incorporate in your work as well: https://github.com/rizinorg/ghidra/commit/5f1f3cf869edda0e596bdb295c1c8c7df68abf96 @ryanmkurtz @caheckman if there is a potential interest in this change I can also send a separate PR with them, just let me know.

RootCubed commented on Ghidra/Features/Decompiler/src/decompile/cpp/printc.cc:
```diff
@@ -3051,6 +3064,37 @@ void PrintC::emitBlockSwitch(const BlockSwitch *bl)
+/// \brief Emits an opening brace according to the indentation style selected in options
+///
+/// \param indent is the current indent level
+int4 PrintC::emitFormattedStartBrace(int4 indent)
```
This function declares a return type but does not return a value. This can cause undefined behaviour, even if the returned value is never used. You should probably add `return indent;` to the end of the function.

jobermayr commented:
Obsoleted by #4914
